src/v/cloud_storage/remote_partition.cc

```diff
@@ -0,0 +1,641 @@
+                vlog(_ctxlog.debug, "initializing new segment reader");
+                _reader = _partition->borrow_reader(
+                  config, _it->first, _it->second);
+            }
+        }
+        vlog(
+          _ctxlog.debug,
+          "maybe_reset_stream completed {} {}",
+          static_cast<bool>(_reader),
+          is_end_of_stream());
+        co_return static_cast<bool>(_reader);
+    }
+
+    /// Transition reader to the completed state. Stop tracking state in
+    /// the 'remote_partition'
+    ss::future<> set_end_of_stream() {
+        co_await _reader->stop();
+        _it = _partition->_segments.end();
+        _reader = {};
+    }
+
+    retry_chain_node _rtc;
+    retry_chain_logger _ctxlog;
+
+    ss::lw_shared_ptr<remote_partition> _partition;
+    /// Currently accessed segment
+    remote_partition::segment_map_t::iterator _it;
+    /// Reader state that was borrowed from the materialized_segment_state
+    std::unique_ptr<remote_segment_batch_reader> _reader;
+    /// Cancelation subscription
+    ss::abort_source::subscription _as_sub;
+};
+
+remote_partition::remote_partition(
+  const manifest& m, remote& api, cache& c, s3::bucket_name bucket)
+  : _rtc()
+  , _ctxlog(cst_log, _rtc, m.get_ntp().path())
+  , _api(api)
+  , _cache(c)
+  , _manifest(m)
+  , _bucket(std::move(bucket))
+  , _stm_jitter(stm_jitter_duration) {}
+
+ss::future<> remote_partition::start() {
+    update_segmnets_incrementally();
+    (void)run_eviction_loop();
+
+    _stm_timer.set_callback([this] {
+        gc_stale_materialized_segments();
+        if (!_gate.is_closed()) {
+            _stm_timer.rearm(_stm_jitter());
+        }
+    });
+    _stm_timer.rearm(_stm_jitter());
+    co_return;
+}
+
+ss::future<> remote_partition::run_eviction_loop() {
+    // Evict readers asynchronously
+    gate_guard g(_gate);
+    while (!_gate.is_closed()) {
+        try {
+            co_await _cvar.wait(
+              [this] { return _gate.is_closed() || !_eviction_list.empty(); });
+        } catch (const ss::broken_condition_variable&) {
+        }
+        auto tmp_list = std::move(_eviction_list);
+        for (auto& rs : tmp_list) {
+            co_await std::visit([](auto&& rs) { return rs->stop(); }, rs);
+        }
+    }
+    vlog(_ctxlog.debug, "remote partition eviction loop stopped");
+}
+
+void remote_partition::start_readahead(
+  remote_partition::segment_map_t::iterator current) {
+    struct visit_hydrate {
+        segment_state& state;
+        remote_partition* part;
+        const model::offset offset_key;
+
+        void operator()(offloaded_segment_ptr& st) {
+            vlog(
+              part->_ctxlog.debug,
+              "remote partition readahead, hydrating {}",
+              manifest_key_to_string(st->manifest_key));
+            auto tmp = st->materialize(*part, offset_key);
+            (void)tmp->segment->hydrate().discard_result().handle_exception(
```

Comment:
this looks like it needs to be in a gate

Reply:
segment has its own gate and we're waiting for all segments to be stopped in `stop` method